### PR TITLE
FIX-#3945: bump omnisci version to 5.10.1

### LINKED ISSUE #3945: Update OmniSci to 5.10.1
OmniSci released a new 5.10.1 version in [conda-forge](https://anaconda.org/conda-forge/pyomniscidbe), we have to bump the version in our testing deps and see how it's going.

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?
This PR bumps OmniSci version to 5.10.1 and PyArrow to 6.* as well (as this is the minimum supported version by OmniSci now).

### Changes in PyArrow 6.0 to align with
Arrow introduced an implicit `time32` values parsing when reading from CSV ([ARROW-11243](https://issues.apache.org/jira/browse/ARROW-11243)), this brought 2 problems for us:
1. We can't easily disable the automatic inference (the same problems as in #3485), that's why we just ignore the differences in dtypes between Modin on OmniSci and pandas in IO-tests (added another exception in [`align_datetime_dtypes`](https://github.com/dchigarev/modin/blob/3738915bdb43fe6308857da8021331ff1ed5644e/modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py#L76) - preprocessor function before asserting equality in IO tests).
2. Pyarrow can't convert `pyarrow.Time32Type` to pandas-supported dtype (`.cast_to_pandas_dtype()` raises a `NotImplementedError`). Added manual conversion of the arrow time type to the python's `datetime.time` at [`PandasDataframe._arrow_type_to_dtype`](https://github.com/dchigarev/modin/blob/3738915bdb43fe6308857da8021331ff1ed5644e/modin/core/dataframe/pandas/dataframe/dataframe.py#L2367). `datetime.time` is the actual type that is stored in a frame when arrow table is converted to pandas:
```python
>>> at = pyarrow.Table.from_pydict({"a": [10, 20]}, schema=pyarrow.schema({"a": pyarrow.time32("s")}))
>>> at.to_pandas()
          a
0  00:00:10
1  00:00:20
>>> at.to_pandas().dtypes
a    object
dtype: object
>>> at.to_pandas().iloc[0, 0]
datetime.time(0, 0, 10)
```

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3945 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py [hunk 1]
```diff
@@ -87,22 +89,46 @@ def align_datetime_dtypes(*dfs):
...
-        # https://github.com/modin-project/modin/issues/3428
-        if len(datetime_cols)
-        else dfs
-    )
+        # so casting it to pandas
+        pandas_df = try_cast_to_pandas(df)
+        if len(datetime_cols):
+            pandas_df = pandas_df.astype(datetime_cols)
+        if len(time_cols):
+            pandas_df = pandas_df[time_cols].applymap(convert_to_time)
```
Comment (ienkovich) [inline modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py:129]: Wouldn't this strip out all non-time columns?

Comment (dchigarev) [reply modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py]: Very nice catch! Fixed it.

**Review by Garra1980 — DISMISSED** [review]
LGTM

**Review by ienkovich — APPROVED** [review]
LGTM!

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 2]
```diff
@@ -2377,7 +2378,16 @@ def _arrow_type_to_dtype(cls, arrow_type):
         object
             Any dtype compatible with pandas.
         """
-        res = arrow_type.to_pandas_dtype()
+        import pyarrow
+
+        try:
+            res = arrow_type.to_pandas_dtype()
+        except NotImplementedError as e:
```
Comment (vnlitvinov) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:2385]: could you please add a comment here explaining why is this needed? this info is now present in PR message only

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py [hunk 3]
```diff
@@ -87,22 +89,48 @@ def align_datetime_dtypes(*dfs):
     -----
     Passed Modin frames may be casted to pandas in the result.
     """
-    datetime_cols = {}
+    datetime_cols = dict()
```
Comment (vnlitvinov) [inline modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py:92]: pointless change IMHO

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py [hunk 4]
```diff
@@ -87,22 +89,48 @@ def align_datetime_dtypes(*dfs):
...
                 datetime_cols[col] = dtype
-
-    casted_dfs = (
+            # datetime.time is considered to be an 'object' dtype in pandas that's why
+            # we have to explicitly check the values type in the column
+            elif (
+                dtype == np.dtype("O")
+                and col not in time_cols
+                # Implying that the data is homogeneous, so checking the first value is enough
+                and (len(df) > 0 and isinstance(df[col].iloc[0], datetime.time))
```
Comment (vnlitvinov) [inline modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py:105]: this could be wrong on so many cases... what if data is `None`-filled at least?

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py [hunk 5]
```diff
@@ -87,22 +89,48 @@ def align_datetime_dtypes(*dfs):
...
+
+    if len(datetime_cols) == 0 and len(time_cols) == 0:
+        return dfs
+
+    def convert_to_time(value):
+        """Convert passed value to `datetime.time`."""
+        if isinstance(value, datetime.time):
+            return value
+        elif isinstance(value, str):
+            return datetime.time(*map(int, value.split(":")))
```
Comment (vnlitvinov) [inline modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py:117]: This feels wrong. There are methods of parsing a string into a date/time/datetime, please use them.

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py [hunk 6]
```diff
@@ -87,22 +89,48 @@ def align_datetime_dtypes(*dfs):
...
-        else dfs
-    )
+        # so casting it to pandas
+        pandas_df = try_cast_to_pandas(df)
+        if len(datetime_cols):
+            pandas_df = pandas_df.astype(datetime_cols)
+        if len(time_cols):
+            pandas_df[list(time_cols)] = pandas_df[list(time_cols)].applymap(
+                convert_to_time
+            )
```
Comment (vnlitvinov) [inline modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py:131]: isn't it going to be suuuuuuuuuuuuuuuuuper slow on big tables?

File: modin/core/dataframe/pandas/dataframe/dataframe.py (hunk 2 above)
Comment (dchigarev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:2387]: added the following comment:
"Conversion to pandas is not implemented for some arrow types, perform manual conversion for them:
"

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py (hunk 5 above)
Comment (dchigarev) [reply modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py]: changed to the `datetime.time.fromisoformat(value)`

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py (hunk 3 above)
Comment (dchigarev) [reply modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py]: reverted this change

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py (hunk 4 above)
Comment (dchigarev) [reply modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py]: added an explicit check for all of the values:
```python
and all(
    isinstance(val, datetime.time) or pandas.isna(val)
    for val in df[col]
)
```
it's going to be slow, but it seems that we have no choise

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py (hunk 6 above)
Comment (dchigarev) [reply modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py:136]: It is, but this logic is only applied in IO tests, I believe we don't read that big files, so the slowness be visible

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 7]
```diff
@@ -2377,7 +2378,18 @@ def _arrow_type_to_dtype(cls, arrow_type):
         object
             Any dtype compatible with pandas.
         """
-        res = arrow_type.to_pandas_dtype()
+        import pyarrow
```
Comment (YarShev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:2381]: Should be placed at the top of this file.

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/partitioning/partition_manager.py [hunk 8]
```diff
@@ -191,14 +191,26 @@ def _get_unsupported_cols(cls, obj):
             else:
                 obj = at
 
+        def is_supported_object_type(dtype):
```
Comment (YarShev) [inline modin/experimental/core/execution/native/implementations/omnisci_on_native/partitioning/partition_manager.py:194]:
```suggestion
        def is_supported_dtype(dtype):
```

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py [hunk 9]
```diff
@@ -88,21 +90,52 @@ def align_datetime_dtypes(*dfs):
...
                 datetime_cols[col] = dtype
-
-    casted_dfs = (
+            # datetime.time is considered to be an 'object' dtype in pandas that's why
+            # we have to explicitly check the values type in the column
+            elif (
+                dtype == np.dtype("O")
+                and col not in time_cols
+                # OmniSci has difficulties with empty frames, so explicitly skip them
+                # https://github.com/modin-project/modin/issues/3428
```
Comment (YarShev) [inline modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py:105]: The issue was closed. Does it still persists?

File: modin/core/dataframe/pandas/dataframe/dataframe.py (hunk 7 above)
Comment (dchigarev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:2381]: We don't want to require `pyarrow` to be installed all the time, we need it only if a user wants to build a dataframe from `pyarrow.Table`

Comment (YarShev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:2381]: makes sense

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py (hunk 9 above)
Comment (dchigarev) [reply modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py:105]: I still encounter problems with empty frames, I've updated the linked issue ([comment](https://github.com/modin-project/modin/issues/3428#issuecomment-1014443549))

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/partitioning/partition_manager.py (hunk 8 above)
Comment (dchigarev) [reply modin/experimental/core/execution/native/implementations/omnisci_on_native/partitioning/partition_manager.py]: applied

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py (hunk 9 above)
Comment (YarShev) [reply modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py:105]: Ok, thanks!

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 10]
```diff
@@ -2377,7 +2378,18 @@ def _arrow_type_to_dtype(cls, arrow_type):
...
+
+        try:
+            res = arrow_type.to_pandas_dtype()
+        # Conversion to pandas is not implemented for some arrow types,
+        # perform manual conversion for them:
+        except NotImplementedError as e:
+            if pyarrow.types.is_time(arrow_type):
+                res = np.dtype(datetime.time)
+            else:
+                raise e
```
Comment (vnlitvinov) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:2391]:
```suggestion
                raise
```
this re-raises the exception keeping the original callstack as if no `except` clause was present, which makes it easier to understand things IMHO

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py [hunk 11]
```diff
@@ -88,21 +90,52 @@ def align_datetime_dtypes(*dfs):
...
-        # This is required so we don't try to cast empty OmniSci frames to pandas:
-        # https://github.com/modin-project/modin/issues/3428
-        if len(datetime_cols)
-        else dfs
-    )
+        # so casting it to pandas
+        pandas_df = try_cast_to_pandas(df)
+        if len(datetime_cols):
+            pandas_df = pandas_df.astype(datetime_cols)
+        if len(time_cols):
```
Comment (vnlitvinov) [inline modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py:133]:
```suggestion
        if datetime_cols:
            pandas_df = pandas_df.astype(datetime_cols)
        if time_cols:
```
reads better, works faster (though I don't think one can measure how much faster - the difference is too small, so main point - reads better)

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py [hunk 12]
```diff
@@ -88,21 +90,52 @@ def align_datetime_dtypes(*dfs):
...
+            pandas_df[list(time_cols)] = pandas_df[list(time_cols)].applymap(
```
Comment (vnlitvinov) [inline modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py:134]: better convert `time_cols` to list once before the loop

File: requirements/env_omnisci.yml [hunk 13]
```diff
@@ -13,7 +13,7 @@ dependencies:
   - pytest-xdist>=2.1.0
   - coverage<5.0
   - pygithub==1.53
-  - pyomniscidbe<=5.8
+  - pyomniscidbe==5.10.0
```
Comment (vnlitvinov) [inline requirements/env_omnisci.yml:16]:
```suggestion
  - pyomniscidbe=5.10
```
this would allow 5.10.0, 5.10.1, etc - I guess this should suit us well

File: requirements/env_omnisci.yml [hunk 14]
```diff
@@ -3,7 +3,7 @@ channels:
   - conda-forge
 dependencies:
   - pandas==1.3.5
-  - pyarrow==5.0.0
+  - pyarrow==6.*
```
Comment (vnlitvinov) [inline requirements/env_omnisci.yml:6]:
```suggestion
  - pyarrow=6
```
or `pyarrow 6.*` - no point of mixing asterisk and equality

**Review by vnlitvinov — COMMENTED** [review]
Mostly good, I would like to change the version pinning before I'm good with merging.

File: requirements/env_omnisci.yml (hunk 14 above)
Comment (YarShev) [reply requirements/env_omnisci.yml:6]: Why do we want to use a fuzzy constraint for pyarrow? OmniSci works with an exact pyarrow version, doesn't it?

File: requirements/env_omnisci.yml (hunk 13 above)
Comment (YarShev) [reply requirements/env_omnisci.yml:16]: Not sure, @Garra1980, @ienkovich, thoughts?

File: requirements/env_omnisci.yml (hunk 14 above)
Comment (Garra1980) [reply requirements/env_omnisci.yml:6]: Omnisci recipe has "set arrow_version = "6.*"" so I think this is fine

File: requirements/env_omnisci.yml (hunk 13 above)
Comment (Garra1980) [reply requirements/env_omnisci.yml:16]: I think yes, =5.10 is ok here

File: modin/core/dataframe/pandas/dataframe/dataframe.py (hunk 10 above)
Comment (YarShev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:2391]: fixed

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py (hunk 11 above)
Comment (YarShev) [reply modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py:133]: fixed

File: modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py (hunk 12 above)
Comment (YarShev) [reply modin/experimental/core/execution/native/implementations/omnisci_on_native/test/utils.py:134]: fixed

File: requirements/env_omnisci.yml (hunk 14 above)
Comment (YarShev) [reply requirements/env_omnisci.yml:6]: fixed

File: requirements/env_omnisci.yml (hunk 13 above)
Comment (YarShev) [reply requirements/env_omnisci.yml:16]: fixed

**Review by devin-petersohn — APPROVED** [review]
LGTM on arrow type conversion. Thanks @dchigarev!

**Review by Garra1980 — APPROVED** [review]
LGTM, thanks a lot everybody!
